libfreerdp/core/gateway/rdg.c

```diff
@@ -669,6 +711,76 @@ static BOOL rdg_process_handshake_response(rdpRdg* rdg, wStream* s)
 	return rdg_send_tunnel_request(rdg);
 }
 
+static BOOL rdg_process_tunnel_response_optional(rdpRdg* rdg, wStream* s, UINT16 fieldsPresent)
+{
+	if (fieldsPresent & HTTP_TUNNEL_RESPONSE_FIELD_TUNNEL_ID)
+	{
+		/* Seek over tunnelId (4 bytes) */
+		if (!Stream_SafeSeek(s, 4))
+		{
+			WLog_ERR(TAG, "[%s] Short packet %" PRIuz ", expected 4", __FUNCTION__,
+			         Stream_GetRemainingLength(s));
+			return FALSE;
+		}
+	}
+
+	if (fieldsPresent & HTTP_TUNNEL_RESPONSE_FIELD_CAPS)
+	{
+		UINT32 caps;
+		if (Stream_GetRemainingLength(s) < 4)
+		{
+			WLog_ERR(TAG, "[%s] Short packet %" PRIuz ", expected 4", __FUNCTION__,
+			         Stream_GetRemainingLength(s));
+			return FALSE;
+		}
+
+		Stream_Read_UINT32(s, caps);
+		WLog_DBG(TAG, "capabilities=%s", capabilities_enum_to_string(caps));
+	}
+
+	if (fieldsPresent & HTTP_TUNNEL_RESPONSE_FIELD_SOH_REQ)
+	{
+		UINT16 certLen;
+
+		/* Seek over nonce (20 bytes) */
+		if (!Stream_SafeSeek(s, 20))
+		{
+			WLog_ERR(TAG, "[%s] Short packet %" PRIuz ", expected 20", __FUNCTION__,
+			         Stream_GetRemainingLength(s));
+			return FALSE;
+		}
+
+		/* Read serverCert */
+		if (!rdg_read_http_unicode_string(s, NULL, NULL))
+		{
+			WLog_ERR(TAG, "[%s] Failed to read string", __FUNCTION__);
+			return FALSE;
+		}
+	}
+
+	if (fieldsPresent & HTTP_TUNNEL_RESPONSE_FIELD_CONSENT_MSG)
+	{
+		WCHAR* msg;
+		UINT16 msgLenBytes;
+		rdpContext* context = rdg->context;
+
+		assert(context);
+		assert(context->instance);
+
+		/* Read message string and invoke callback */
+		if (!rdg_read_http_unicode_string(s, &msg, &msgLenBytes))
```

Comment:
Is this string really `\0` terminated? (unchecked data from server)

Reply:
The documentation does not say anything about \0 termination but I check if there are at least msgLenBytes of string data available and the both parameters (buffer and buffer size) are forwarded to the callback